File: src/libs/cpp/functions/explode.cpp
#include "explode.hpp"

#include <sstream>

void explode(Context &context) noexcept
{
	const auto arg = context.stack.pop();
	std::vector<Value> params;

	if (std::holds_alternative<std::string>(arg))
	{
		// Split strings by newline
		std::stringstream stream(std::get<std::string>(arg));
		std::string line;
		while (std::getline(stream, line))
		{
			params.push_back(line);
		}
	}
	else
	{
		params = arg.to_array();
	}

	for (size_t i = params.size(); i > 0; i--)
	{
		context.stack.push(params[i - 1]);
	}
}

File: src/runtime/functions/explode.lua
return function(vm, line, param)
	local array = vm.pop()
	local tp = std.type(array)

	if tp == 'object' then
		for key, value in pairs(array) do vm.push(key) end
	elseif tp == 'array' then
		for i = #array, 1, -1 do
			vm.push(array[i])
		end
	elseif tp == 'string' then
		array = std.split(array, '\n')
		for i = #array, 1, -1 do
			vm.push(array[i])
		end
	else
		vm.push(array)
	end
end
